In [16]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

In [4]:
train_df = pd.read_csv(Path('Resources/Generator/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/Generator/2020Q1loans.csv'))

Initial Prediction: Looking at the two model options, the random forest classifier should perform better based upon the mechanics behind the algorithm. This is due to the trees that the algorithm uses to determine how high risk the loans are.

In [7]:
# Convert categorical data to numeric and separate target feature for training data
x_train = pd.get_dummies(train_df.drop(columns=['target']))
y_train = train_df['target']

In [14]:
# Convert categorical data to numeric and separate target feature for testing data
x_test = pd.get_dummies(test_df.drop(columns=['target']))
y_test = test_df['target']

In [15]:
# add missing dummy variables to testing set
for col in x_train.columns:
    if col not in x_test.columns:
        x_test[col]=0

In [18]:
# Train the Logistic Regression model on the unscaled data and print the model score
classifier = LogisticRegression()
classifier.fit(x_train, y_train)
classifier.score(x_test, y_test)

C:\Users\14043\Downloads\Newfolder\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.5074436410038281

In [19]:
# Train a Random Forest Classifier model and print the model score
rfc = RandomForestClassifier()
rfc.fit(x_train, y_train)
rfc.score(x_test, y_test)

0.6399404508719694

In [21]:
# Scale the data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(x_train)
x_train_scaled = scaler.transform(x_train)
x_test_scaled = scaler.transform (x_test)

In [22]:
# Train the Logistic Regression model on the scaled data and print the model score
classifier = LogisticRegression()
classifier.fit(x_train_scaled, y_train)
classifier.score(x_test_scaled, y_test)

C:\Users\14043\Downloads\Newfolder\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.7598894087622289

In [23]:
# Train a Random Forest Classifier model on the scaled data and print the model score
rfc = RandomForestClassifier()
rfc.fit(x_train_scaled, y_train)
rfc.score(x_test_scaled, y_test)

0.6363249680986814

Final Outcome: After reviewing the outcomes from our regression models we can see that the logistic regression model performed better on scaled data at 76% while the random forest classifier stayed constant regardless of scaling at ~63%. This means that using the logistic regression model would actually be better suited to use for classsification of high risk loans vs. the low risk loans.